src/hotspot/cpu/x86/stubGenerator_x86_64_poly.cpp

```diff
@@ -1023,3 +1048,497 @@ address StubGenerator::generate_poly1305_processBlocks() {
+  __ vpsrlq(YTMP1, P0L, 44, Assembler::AVX_256bit);
+  __ vpsllq(P0H, P0H, 8, Assembler::AVX_256bit);
+  //
+  __ vpmadd52luq(P1L, A1, R0, Assembler::AVX_256bit);
+  __ vpmadd52huq(P1H, A1, R0, Assembler::AVX_256bit);
+  // Carry propagation (first pass) - continue
+  __ vpand(A0, P0L, ExternalAddress(poly1305_mask44()), Assembler::AVX_256bit, rscratch); // Clear top 20 bits
+  __ vpaddq(P0H, P0H, YTMP1, Assembler::AVX_256bit);
+  //
+  __ vpmadd52luq(P2L, A1, R1, Assembler::AVX_256bit);
+  __ vpmadd52huq(P2H, A1, R1, Assembler::AVX_256bit);
+
+  // Carry propagation (first pass) - continue
+  __ vpaddq(P1L, P1L, P0H, Assembler::AVX_256bit);
+  __ vpsllq(P1H, P1H, 8, Assembler::AVX_256bit);
+  __ vpsrlq(YTMP1, P1L, 44, Assembler::AVX_256bit);
+  __ vpand(A1, P1L, ExternalAddress(poly1305_mask44()), Assembler::AVX_256bit, rscratch); // Clear top 20 bits
+
+  __ vpaddq(P2L, P2L, P1H, Assembler::AVX_256bit);
+  __ vpaddq(P2L, P2L, YTMP1, Assembler::AVX_256bit);
+  __ vpand(A2, P2L, ExternalAddress(poly1305_mask42()), Assembler::AVX_256bit, rscratch); // Clear top 22 bits
+  __ vpsrlq(YTMP1, P2L, 42, Assembler::AVX_256bit);
+  __ vpsllq(P2H, P2H, 10, Assembler::AVX_256bit);
+  __ vpaddq(P2H, P2H, YTMP1, Assembler::AVX_256bit);
+
+  // Carry propagation (second pass)
+  // Multiply by 5 the highest bits (above 130 bits)
+  __ vpaddq(A0, A0, P2H, Assembler::AVX_256bit);
+  __ vpsllq(P2H, P2H, 2, Assembler::AVX_256bit);
+  __ vpaddq(A0, A0, P2H, Assembler::AVX_256bit);
+
+  __ vpsrlq(YTMP1, A0, 44, Assembler::AVX_256bit);
+  __ vpand(A0, A0, ExternalAddress(poly1305_mask44()), Assembler::AVX_256bit, rscratch); // Clear top 20 bits
+  __ vpaddq(A1, A1, YTMP1, Assembler::AVX_256bit);
+
+}
+
+void StubGenerator::poly1305_msg_mul_reduce_vec4_avx2(
+  const XMMRegister A0, const XMMRegister A1, const XMMRegister A2,
+  const Address R0, const Address R1, const Address R2,
+  const Address R1P, const Address R2P,
+  const XMMRegister P0L, const XMMRegister P0H,
+  const XMMRegister P1L, const XMMRegister P1H,
+  const XMMRegister P2L, const XMMRegister P2H,
+  const XMMRegister YTMP1, const XMMRegister YTMP2,
+  const XMMRegister YTMP3, const XMMRegister YTMP4,
+  const XMMRegister YTMP5, const XMMRegister YTMP6,
+  const Register input, const Register length, const Register rscratch)
+{
+  // Reset accumulator
+  __ vpxor(P0L, P0L, P0L, Assembler::AVX_256bit);
+  __ vpxor(P0H, P0H, P0H, Assembler::AVX_256bit);
+  __ vpxor(P1L, P1L, P1L, Assembler::AVX_256bit);
+  __ vpxor(P1H, P1H, P1H, Assembler::AVX_256bit);
+  __ vpxor(P2L, P2L, P2L, Assembler::AVX_256bit);
+  __ vpxor(P2H, P2H, P2H, Assembler::AVX_256bit);
+
+  // Calculate partial products
+  // p0 = a2xr1'
+  // p1 = a2xr2'
+  // p2 = a2xr0
+      __ vpmadd52luq(P0L, A2, R1P, Assembler::AVX_256bit);
+      __ vpmadd52huq(P0H, A2, R1P, Assembler::AVX_256bit);
+  // Interleave input loading with hash computation
+  __ lea(input, Address(input,16*4));
+  __ subl(length, 16*4);
```

Comment:
This really belongs in the caller, next to the labels and compare

Reply:
This is how Pablo implemented it.